Add specification for WebSocket (#413)
ballerina-platform/module-ballerina-websocket: docs/spec/spec.md

# Specification: Ballerina WebSocket Library

> _Owners_: @shafreenAnfar @bhashinee

Reviewer: ```suggestion
_Owners_: @shafreenAnfar @bhashinee  
```

_Reviewers_: @shafreenAnfar    
_Created_: 2021/12/09  
_Updated_: 2021/12/09  
_Issue_: [#2165](https://github.com/ballerina-platform/ballerina-standard-library/issues/2165)

# Introduction

> This is the specification for WebSocket standard library which is used to implement WebSocket compliant 'listener' and 'client' using [Ballerina programming language](https://ballerina.io/), which is an open-source programming language for the

Reviewer: ```suggestion
This is the specification for WebSocket standard library which is used to implement WebSocket compliant `listener` and `client` using [Ballerina programming language](https://ballerina.io/), which is an open-source programming language for the
```

cloud that makes it easier to use, combine, and create network services. 

# Contents
1. [Overview](#1-overview)
2. [Listener](#2-listener)
3. [Service Types](#3-service-types)
   * 3.1. [Upgrade Service](#31-upgrade-service)
   * 3.2. [WebSocket Service](#32-websocket-service)
4. [Client](#4-client)
5. [Securing the WebSocket Connections](#5-securing-the-websocket-connections)
   * 5.1. [SSL/TLS](#51-ssl-tls)
   * 5.2. [Authentication and Authorization](#52-authentication-and-authorization)
6. [Samples](#6-samples)

## 1. [Overview](#1-overview)

WebSocket is a protocol that allows a long held full-duplex connection between a server and client. This specification elaborates on how Ballerina language provides a tested WebSocket client and server implementation that is compliant with the [RFC 6455](https://www.rfc-editor.org/rfc/rfc6455.html).

## 2. [Listener](#2-listener)

The WebSocket listener can be constructed with a port or an `http:Listener`. When initiating the listener it opens up the port and attaches the upgrade service which is quite similar to an HTTP service at the given service path. 

> 

Reviewer: The WebSocket listener can be constructed with a port or an `http:Listener`. When initiating the listener it opens up the port and attaches the upgrade service at the given service path. It is also worth noting that upgrade service is quite similar to an HTTP service.

> 

Reviewer: ```suggestion
The WebSocket listener can be constructed with a port or an http:Listener. When initiating the listener it opens up the port and attaches the upgrade service at the given service path. It is also worth noting that upgrade service is quite similar to an HTTP service.
```

## 3. [Service Types](#3-service-types)

### 3.1 UpgradeService

Upgrade service is pretty much similar to an HTTP service. It has a single `get` resource, which takes in an `http:Request` optionally. The `get` resource returns a `websocket:Service` to which incoming messages get dispatched after a successful WebSocket connection upgrade. This resource can be used to intercept the initial HTTP upgrade with custom headers or to cancel the WebSocket upgrade by returning an error.

### 3.1 Service

> Once the WebSocket upgrade is accepted by the UpgradeService, it returns a `ws:Service`. This service has a fixed set of remote functions(onTextMessage, onBinaryMessage, onError, onOpen, onIdleTimeout, onClose, onPing, onPong) that do not have any configs. Receiving messages will get dispatched to the relevant remote function. 

Reviewer: ```suggestion
Once the WebSocket upgrade is accepted by the UpgradeService, it returns a `websocket:Service `. This service has a fixed set of remote functions(onTextMessage, onBinaryMessage, onError, onOpen, onIdleTimeout, onClose, onPing, onPong) that do not have any configs. Receiving messages will get dispatched to the relevant remote function. 
```

```ballerina
service /ws on new websocket:Listener(21003) {
    resource function get .(http:Request req) returns websocket:Service|websocket:UpgradeError {
        return new WsService();
}
        
service class WsService {
  *websocket:Service;
  remote isolated function onTextMessage(websocket:Caller caller, string data) returns websocket:Error? {
      check caller->writeTextMessage(data);
  }

> }              

Reviewer: Formatting is inconsistent.

```

#### Remote methods associated with websocket:Service

> 

Reviewer: Let's make subtopics for each remote method

    io:println(string `Ping received with data: ${data.toBase64()}`);
    check caller->pong(data);
}
 
remote function onPong(websocket:Caller caller, byte[] data) {
    io:println(string `Pong received with data: ${data.toBase64()}`);
}
```

onIdleTimeout: This remote method is dispatched when the idle timeout is reached. The idleTimeout has to be configured either in the WebSocket service or the client configuration.
```ballerina
remote function onIdleTimeout(websocket:Client caller) {
    io:println("Connection timed out");
}
```

onClose: This remote method is dispatched when a close frame with a statusCode and a reason is received.
```ballerina
remote function onClose(websocket:Caller caller, int statusCode, string reason) {
    io:println(string `Client closed connection with ${statusCode} because of ${reason}`);
}
```

onError: This remote method is dispatched when an error occurs in the WebSocket connection. This will always be preceded by a connection closure with an appropriate close frame.
```ballerina
remote function onError(websocket:Caller caller, error err) {
    io:println(err.message());
}
```

## 4. [Client](#4-client)

> `websocket:Client` can be used to send and receive data synchronously over WebSocket connection. 

Reviewer: Let's mention that the underline implementation is non-blocking.

#### Send and receive messages using 'websocket:Client'

> 

Reviewer: Let's have a sub section for each remote service. Each subsection we can have signature of each remote method and the explain it.

1. Send a text message.
```ballerina
   check wsClient->writeTextMessage("Text message");
```

2. Send a binary message
```ballerina
   check wsClient->writeBinaryMessage("Text message".toBytes());
```

3. Receive a text message
```ballerina
   string textResp = check wsClient->readTextMessage();
```

4. Receive a binary message
```ballerina
   byte[] textResp = check wsClient->readBinaryMessage();
```

5. Receive a message without prior knowledge of message type.
```ballerina
   byte[]|string|websocket:Error data = wsClient->readMessage();
   if (data is string) {
       io:println(data);
   } else if (data is byte[]) {
       io:println(data);
   } else {
       io:println("Error occurred", data.message());
   }
```
6. Send ping message
```ballerina
   check wsClient->ping([5, 24, 56, 243]);
```

7. Send pong message
```ballerina
   check wsClient->pong([5, 24, 56, 243]);
```

> 8. Receive ping/pong messages

Reviewer: Here also let's include the signature of of onPing and onPong

To receive ping/pong messages, users have to register a `websocket:PingPongService` when creating the client. If the service is registered, receiving ping/pong messages will get dispatched to the `onPing` and `onPong` remote functions respectively.
```ballerina
   service class PingPongService {
       *websocket:PingPongService;
       remote function onPong(websocket:Caller wsEp, byte[] data) {
           io:println("Pong received", data);
       }
       
       remote isolated function onPing(websocket:Caller caller, byte[] localData) returns byte[] {
           return localData;
       }    
       
   }
   
   websocket:Client wsClient = check new ("ws://localhost:21020", {pingPongHandler : new PingPongService()});
```
If the user has implemented `onPing` on their service, it's user's responsibility to send the `pong` frame. It can be done simply by returning the data from the remote function, or else can be done using the `pong` API of websocket:Caller. If the user hasn't implemented the `onPing` remote function, `pong` will be sent automatically.

> 9. Close the connection

Reviewer: This has to come before explaining ping/pong

# Specification: Ballerina WebSocket Library

_Owners_: @shafreenAnfar @bhashinee 

> _Reviewers_: @shafreenAnfar    

Reviewer: It seems formatting issue is still there.

_Created_: 2021/12/09  
_Updated_: 2021/12/09  
_Issue_: [#2165](https://github.com/ballerina-platform/ballerina-standard-library/issues/2165)

# Introduction

This is the specification for WebSocket standard library which is used to implement WebSocket compliant `listener` and `client` using [Ballerina programming language](https://ballerina.io/), which is an open-source programming language for the
cloud that makes it easier to use, combine, and create network services. 

# Contents
1. [Overview](#1-overview)
2. [Listener](#2-listener)
3. [Service Types](#3-service-types)
   * 3.1. [Upgrade Service](#31-upgrade-service)
   * 3.2. [WebSocket Service](#32-websocket-service)

> 4. [Client](#4-client)

Reviewer: We need to update the table content with the remote methods.

6. [Samples](#6-samples)

## 1. [Overview](#1-overview)

WebSocket is a protocol that allows a long held full-duplex connection between a server and client. This specification elaborates on how Ballerina language provides a tested WebSocket client and server implementation that is compliant with the [RFC 6455](https://www.rfc-editor.org/rfc/rfc6455.html).

## 2. [Listener](#2-listener)

The WebSocket listener can be constructed with a port or an http:Listener. When initiating the listener it opens up the port and attaches the upgrade service at the given service path. It is also worth noting that upgrade service is quite similar to an HTTP service.
## 3. [Service Types](#3-service-types)

### 3.1 UpgradeService

Upgrade service is pretty much similar to an HTTP service. It has a single `get` resource, which takes in an `http:Request` optionally. The `get` resource returns a `websocket:Service` to which incoming messages get dispatched after a successful WebSocket connection upgrade. This resource can be used to intercept the initial HTTP upgrade with custom headers or to cancel the WebSocket upgrade by returning an error.

### 3.2 Service

> Once the WebSocket upgrade is accepted by the UpgradeService, it returns a `websocket:Service`. This service has a fixed set of remote functions(onTextMessage, onBinaryMessage, onError, onOpen, onIdleTimeout, onClose, onPing, onPong) that do not have any configs. Receiving messages will get dispatched to the relevant remote function. 

Reviewer: Instead of adding the remote methods as a long list. We can simply say something as follows,

This service has a fixed set of remote functions that do not have any configs. Receiving messages will get dispatched to the relevant remote function. Each remote function is explained below.

```ballerina
service /ws on new websocket:Listener(21003) {
    resource function get .(http:Request req) returns websocket:Service|websocket:UpgradeError {
        return new WsService();
    }    
}
        
service class WsService {
    *websocket:Service;
    remote isolated function onTextMessage(websocket:Caller caller, string data) returns websocket:Error? {
        check caller->writeTextMessage(data);
    }
}              
```

#### Remote methods associated with websocket:Service

> ##### onOpen

Reviewer: We need to update table of content for these sub remote methods.
